src/qibolab/platform/platform.py

```diff
@@ -214,60 +214,66 @@ def disconnect(self):
                 instrument.disconnect()
         self.is_connected = False
 
-    def _execute(self, sequence, options, **kwargs):
+    @property
+    def _controller(self):
+        """Identify controller instrument.
+
+        Used for splitting the unrolled sequences to batches.
+
+        This method does not support platforms with more than one
+        controller instruments.
+        """
+        controllers = [
+            instr
+            for instr in self.instruments.values()
+            if isinstance(instr, Controller)
+        ]
+        assert len(controllers) == 1
+        return controllers[0]
+
+    def _execute(self, sequence, options, *sweepers):
         """Executes sequence on the controllers."""
         result = {}
 
         for instrument in self.instruments.values():
             if isinstance(instrument, Controller):
                 new_result = instrument.play(
-                    self.qubits, self.couplers, sequence, options
+                    self.qubits, self.couplers, sequence, options, *sweepers
                 )
                 if isinstance(new_result, dict):
                     result.update(new_result)
 
         return result
 
-    def execute_pulse_sequence(
-        self, sequence: PulseSequence, options: ExecutionParameters, **kwargs
-    ):
-        """
-        Args:
-            sequence (:class:`qibolab.pulses.PulseSequence`): Pulse sequences to execute.
-            options (:class:`qibolab.platforms.platform.ExecutionParameters`): Object holding the execution options.
-            **kwargs: May need them for something
-        Returns:
-            Readout results acquired by after execution.
-        """
-        options = self.settings.fill(options)
+    def execute(
+        self,
+        sequences: List[PulseSequence],
+        options: ExecutionParameters,
+        *sweepers: Sweeper,
+    ) -> dict[Any, list]:
+        """Execute a pulse sequences.
```

Comment:
Maybe in this docstring, most likely in the `Returns:` part (but I cannot comment directly there), we could add a more detailed sentence about what exactly is returned. For example a `dict` from readout pulse `id`s to `ExecuteResult` objects.

That reminds me that at some point we need to finalize the output of this function for 0.2, since this is also user facing. Certainly not for this PR, but currently, I am not sure about the following:
* Should we keep the `dict` and with what keys? Currently we are using `serial` (replaced by `id`) and also qubit name. Personally, I think we should keep the `dict` but only with `id`s, not qubits.
* Should we keep separate objects for different averaging/acquisition modes, or just replace everything with array?
* Unrolling: I think `id` is still sufficient and we don't need lists, if all readouts in the given `sequences` have distinct `id`s.
* Sweepers: I think here the same `id` is associated to multiple measurements so maybe we need multi-dimensional arrays (as we have now), however I could think some complicated cases (eg. we are sweeping something on the readout pulse).

Reply:
> Should we keep the `dict` and with what keys? Currently we are using `serial` (replaced by `id`) and also qubit name. Personally, I think we should keep the `dict` but only with `id`s, not qubits.

Agreed

> Should we keep separate objects for different averaging/acquisition modes, or just replace everything with array?

Replace with arrays. #809 #899 are both listed for 0.2

> Unrolling: I think `id` is still sufficient and we don't need lists, if all readouts in the given `sequences` have distinct `id`s.

That is fine by me. It would keep the return object flat (less nested)

> Sweepers: I think here the same `id` is associated to multiple measurements so maybe we need multi-dimensional arrays (as we have now), however I could think some complicated cases (eg. we are sweeping something on the readout pulse).

Sweepers are for loops, by definition. We should collect them within multidimensional arrays, since each iteration is homogeneous with the others.
Possibly, we should identify results with just the id of the original pulse, if newer readout pulses are introduced in the process. But this should be fully internal to the drivers, despite non-trivial